radix-engine/src/track/db_key_mapper.rs

```diff
@@ -1,71 +1,147 @@
+    /// you are most likely using the "partition vs sort key" construct in a wrong way).
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey;
 
     /// Converts the given Substate key to the database's sort key.
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey;
+
+    /// Converts the given database's sort key to a Substate key.
+    fn from_db_sort_key(db_sort_key: &DbSortKey) -> SubstateKey;
 }
 
 /// A [`DatabaseKeyMapper`] tailored for databases which cannot tolerate long common prefixes
 /// among keys (for performance reasons). In other words, it spreads the keys "evenly" (i.e.
-/// pseudo-randomly) across the key space.
-/// For context: our actual use-case for it is the Jellyfish Merkle Tree.
+/// pseudo-randomly) across the key space. For context: our use-case for this is the Jellyfish
+/// Merkle Tree.
+///
+/// This implementation is the actual, protocol-enforced one, to be used in public Radix networks.
+///
+/// This implementation achieves the prefix-spreading by:
+/// - using a (long, ~100% unique) hash instead of plain RE Node and Module ID (please note that it
+///   makes this mapping effectively irreversible);
+/// - using a (shorter, but hard to crack) hash prefix for Substate key.
 pub struct SpreadPrefixKeyMapper;
 
 impl DatabaseKeyMapper for SpreadPrefixKeyMapper {
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey {
         let mut buffer = Vec::new();
         buffer.extend(node_id.as_ref());
         buffer.push(module_id.0);
-        let hash_bytes = hash(buffer).0[(Hash::LENGTH - 26)..Hash::LENGTH].to_vec(); // 26 bytes
+        let hash_bytes = hash(buffer).0[..Self::PARTITION_KEY_HASH_LENGTH].to_vec();
         DbPartitionKey(hash_bytes)
     }
 
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey {
-        let bytes = match key {
-            SubstateKey::Tuple(field) => {
-                vec![*field]
-            }
-            SubstateKey::Map(key) => {
-                hash(key).0[12..Hash::LENGTH].to_vec() // 20 bytes
-            }
-            SubstateKey::Sorted((bucket, key)) => {
-                let mut bytes = bucket.to_be_bytes().to_vec();
-                bytes.extend(hash(key).0[12..Hash::LENGTH].to_vec()); // 20 bytes
-                bytes
-            }
+        let (discriminator_byte, rest_bytes) = match key {
+            SubstateKey::Tuple(field) => (Self::TUPLE_DISCRIMINATOR, vec![*field]),
+            SubstateKey::Map(key) => (Self::MAP_DISCRIMINATOR, Self::to_hash_prefixed(key)),
+            SubstateKey::Sorted((bucket, key)) => (
+                Self::SORTED_DISCRIMINATOR,
+                [
+                    bucket.to_be_bytes().as_slice(),
+                    &Self::to_hash_prefixed(key),
+                ]
+                .concat(),
+            ),
         };
-        DbSortKey(bytes)
+        DbSortKey(combine(discriminator_byte, &rest_bytes))
```

Comment:
Putting the discriminator byte first feels wrong for JMT performance, and feels weird/wasteful for the "sorted key" abstraction (even if it works) - maybe we could put it at the end?

As it is fixed for a given partition anyway, putting at the end makes most sense I think. It enables invertibility without getting in the way at all.

Reply:
> As it is fixed for a given partition anyway (...)

wowowow, I was not aware of this!
In such case, maybe we should actually have API like `list<K>(NodeId, ModuleId) -> Iterator<K>` (where `K` can be `TupleKey`, `MapKey`, `SortedU16Key`) ?
(no discriminator byte)

If the above is not feasible, then I can definitely put the discriminator at the end 👍